caffe2/contrib/tensorrt/tensorrt_tranformer.cc

```diff
@@ -0,0 +1,423 @@
+    } else {
+      op.add_output(name);
+    }
+  }
+
+  auto engine_plan = InferObject(trt_engine->serialize());
+
+  auto* serialized_engine_arg = op.add_arg();
+  serialized_engine_arg->set_s("");
+  serialized_engine_arg->set_name("serialized_engine");
+  auto* s = serialized_engine_arg->mutable_s();
+  s->assign((char*)engine_plan->data(), engine_plan->size());
+
+  auto* max_batch_size_arg = op.add_arg();
+  max_batch_size_arg->set_name("max_batch_size");
+  max_batch_size_arg->set_i(max_batch_size_);
+
+  auto* verbosity_arg = op.add_arg();
+  verbosity_arg->set_name("log_verbosity");
+  verbosity_arg->set_i(verbosity_);
+
+  auto* output_size_hints_arg = op.add_arg();
+  auto* output_size_names_arg = op.add_arg();
+  output_size_hints_arg->set_name("output_size_hints");
+  output_size_names_arg->set_name("output_size_names");
+  for(const auto& o: op.output()) {
+    const auto it = output_size_hints.find(o);
+    if (it != output_size_hints.end()) {
+      const auto& dims = it->second;
+      for (const auto& i : dims) {
+        output_size_hints_arg->add_ints(i);
+      }
+      // Add an extra -1 to indicate the end
+      output_size_hints_arg->add_ints(-1);
+      output_size_names_arg->add_strings(o);
+      LOG(INFO) << "Adding output hint: " << o;
+    }
+  }
+
+  return op;
+}
+
+void TensorRTTransformer::ClusterToTrtOp(
+    const NetDef& init_net,
+    const NetDef& pred_net,
+    int start,
+    int end,
+    const std::unordered_set<std::string>& weights,
+    const std::unordered_map<std::string, TensorShape>& shape_hints,
+    ::ONNX_NAMESPACE::ModelProto* model,
+    std::vector<OperatorDef>* new_ops) {
+  if (model->graph().node_size() == 0) {
+    return;
+  }
+  model->mutable_graph()->clear_input();
+  model->mutable_graph()->clear_output();
+  model->mutable_graph()->clear_initializer();
+
+  // Figure out the boundary outputs
+  auto outputs = ConvertToValueInfo(FigureOutputs(pred_net, start, end), shape_hints);
+  std::unordered_map<std::string, std::vector<int>> output_shape_hints;
+  for (const auto& i: outputs) {
+    model->mutable_graph()->add_output()->CopyFrom(i);
+    auto ret = output_shape_hints.emplace(i.name(), std::vector<int>());
+    auto& vec = ret.first->second;
+    const auto it = shape_hints.find(i.name());
+    CAFFE_ENFORCE(it != shape_hints.end(), "Cannot find shape info for output ", i.name());
+    const auto& shape = it->second;
+    for (int k = 0; k < shape.dims().size(); ++k) {
+      vec.push_back(shape.dims(k));
+    }
+  }
+
+  // Figure out the boundary inputs
+  std::unordered_set<std::string> initialization_list;
+  auto total_inputs_vec = FigureInputs(pred_net, start, end, *new_ops, weights, &initialization_list);
+  auto inputs = ConvertToValueInfo(total_inputs_vec, shape_hints);
+  for (const auto& i: inputs) {
+    LOG(INFO) << "Added input: " << i.name();
+    model->mutable_graph()->add_input()->CopyFrom(i);
+  }
+
+  // Convert weights to initializing tensors
```

Comment:
This is going to be slow. Does TRT support any alternative?

Reply:
Yes. Right now it is. But I think we can extend onnx-trt to support the initializeGraphIO once the backend spec is merged.

Reply:
Sorry not initializeGraphIO but some alternatives. We do want to save copies from c2 to onnx to try.